### DISCUSSION TITLE
Exchange rate in activities

### DISCUSSION THREAD

**Discussion by helgehatt** [ghostfolio/ghostfolio#3141] [Ideas]
I know there are a couple of issues regarding exchange rates (#2700, #2449) and that there has been some work done related to currency conversion (#3039). However, as long as it is not possible to specify the exact exchange rate used in activities in foreign currencies, the investment value in your base currency will always be incorrect. This is because the exchange rate you get when buying shares varies across the whole day, and will never be exactly equal the exchange rate we fetch from Yahoo or wherever. My suggestion is that when you add an activity with a foreign currency, an additional input field with today's "default" exchange rate appears, which you can optionally modify to the exact exchange rate you actually paid.

**Comment by MRobi1** [comment]
I don't believe this is quite the root of the issue.

@dtslvr may be able to provide some additional insight here on exactly how the data is stored, but from what I've been able to determine when you add the activity it converts the transaction using the exchange for that day and stores in the converted currency. Then the performance dashboard converts it back to your selected currency using the _current_ exchange rate. And the difference in exchange rate from the day you made the transaction to the day you're viewing the transaction is what causes the inaccuracy.

I feel there are 2 possible solutions here.
A) Similar to what you're saying, store the activity along with the exchange rate at the time of the transaction and convert each individual activity back to the selected currency using each activity's individual exchange rate. I don't think you need to have the manual aspect here.
B) Skip the currency conversion completely for the purchase price. Just store the purchase data in the selected currency. If I buy for $1,000CAD, there's no sense in converting it to USD then converting it back to CAD at a different exchange rate. Just keep it as $1,000CAD. Then do a currency conversion on only the current value and you've got an accurate performance dashboard. This seems much simpler to me.

My example [here](https://github.com/ghostfolio/ghostfolio/issues/2449#issuecomment-1962987063) shows the issue.

It basically makes ghostfolio completely unusable for any user who has any sort of currency exchange and cares about having an accurate representation of their portfolio's performance. I keep checking back every few weeks in hopes it will someday be fixed and usable again. And until then I try to bring as much awareness as I can to the issue because I'm sure there are many who are using it without even noticing the issue.

**Comment by helgehatt** [comment]
@MRobi1 I have read your bug report and I agree that’s an issue as well, but in my case I’m actually buying foreign currency at the same time I’m buying shares in that foreign currency, so it’s not exactly the same problem.

**Comment by MRobi1** [comment]
> I’m actually buying foreign currency at the same time I’m buying shares in that foreign currency

If it's not the same issue, I'm unclear of exactly what you're describing here. Could you provide a bit more of an example?

In my example, if I'm buying $1,000CAD in AAPL which is traded in USD on the NYSE, the trade is settled in USD. So what the brokerage does is take the $1,000 CAD, exchanges to USD and then the shares are purchased. But my initial investment of $1,000CAD does not change regardless of what the current exchange rate was at the time of purchase or what the exchange rate was is today. $1,000 == $1,000. What is key in the performance reporting is that my initial investment stays at $1,000. This is what is not happening because ghostfolio does the currency exchange to USD, adds all purchases up in USD, then uses the current exchange rate (instead of at the time of purchase) to display the performance calculation. This is why $1,000 becomes $1,018 or $960, or whatever it happens to be at the time of viewing.

This is why either storing it in the native currency being used for performance calculation, or storing the exchange rate ghostfolio used when storing in the database and then converting each transaction back to native currency individually before adding them together would be the right solution. Manually inputting the exact exchange rate shouldn't be needed and often isn't known.

**Comment by helgehatt** [comment]
> Manually inputting the exact exchange rate shouldn't be needed and often isn't known.

Well that is just a false assumption. In my case I buy 100 shares of ASDF at a maximum price of €10 each, so let’s say the total becomes €995. I pay in £, because I don’t have bank accounts in other currencies like € or $. Then the broker which is also my bank provides me with the exchange rate of say 0.85 based on the current market price, which results in total investment of £845.75. That’s the normal procedure where I am from, and requires manually inputting the exact exchange rate to be correct.

**Comment by Sigfrodr** [comment]
Agreed, this is a blocker when dealing with foreign instruments and wanting an accurate view of the position.

**Comment by pmagaz** [comment]
I believe several issues affect currency conversion and activities that, while not identical, are interconnected.

1) Exchange rates:

When using exchange rates from Yahoo or other similar data providers, you only get the closing price, not the rate at which you actually exchanged. On August 2nd 2024 (just a few days ago), USD/CHF dropped 1.78% from opening to closing. This deviation and others will lead to inaccuracies, yes, but I don't see any easy solution for this. Even if you get the fine grained price of 1m bar (it won't be from any free service like Yahoo that for sure), currency exchanges are decentralized so prices, spreads, etc may slighly change from one broker to another. A pain to live with.

2) Automatic Currency Conversion when adding activities:

This is the real problem and the deal breaker to me and many users. Activity Currency is **tied to the traded asset**, not the account/platform/portfolio currency, which surprisingly, has almost zero effect in the app.  If you manually enter an activity in EUR for an asset traded in USD it will automatically converted as already shown by some users. So , activities of a Chinese company traded in Yuans will be converted and shown in the app in Yuans, Euro traded assets in Euros, etc.

Mixing currencies in a portfolio is messy and difficult to justify. People usually operate in their home currency, relying on their broker for currency exchange in a transparent manner for the uer. Auto-converting to the asset's currency **corrupts stored data** (because issue 1) and makes original pricing recovery impossible. This also makes data importing technically impossible. My broker's transactions are in EUR, but when I try to load a CSV, I get an error `(activities.0.currency ("EUR") does not match with currency of TICKER ("USD"))` because the system expects the asset's currency. Expecting users to manually get the daily exchange rate and convert or add trought the app thousands of activities doesnt seem to be reasonable.

I strongly believe that @MRobi1's Option B is the correct/simple approach. To keep data accuracy, activities should be stored in their original currency and converted on the fly to the user's selected currency at account or platform level. This would allow users to view their dashboards, activities, etc in any currency they prefer as in almost every financial app out there. 

As several users have pointed out, at this point, Ghostfolio is unusable. The activities issue seems related to wrong design decisions. While such problems might be "tolerable" in a personal open-source project, they are indefensible in a paid premium offering. I hope this feedback helps the maintainers fix it as soon as possible.

Best
